design/cluster-api-provider-metal3/node_reuse.md

```diff
@@ -0,0 +1,278 @@
+
+## Proposal
+
+We would like to propose an interface to mark BMH(s) that we want to reuse after
+deprovisioning and second, add a selection/filtering mechanism in CAPM3 Machine
+controller that will pick those BMH(s) with a specific label (to be exact, based
+on the label `infrastructure.cluster.x-k8s.io/node-reuse`).
+
+## Design Details
+
+In case of nodeReuse field is set to True, to make sure that a BMH(s) is not
+utilized by other M3M(s) after it has been deprovisioned, the new label
+`infrastructure.cluster.x-k8s.io/node-reuse` will store the name of KCP/MD even
+when it is in `Ready` state. During the upgrade, after KCP/MD creates new
+Machines, CAPM3 will create new M3M(s) and will filter out BMH(s) that have the
+name of KCP/MD in the `infrastructure.cluster.x-k8s.io/node-reuse` label.
+If the name of the KCP/MD matches the name set in
+`infrastructure.cluster.x-k8s.io/node-reuse` label of BMH, then CAPM3 Machine
+controller will associate that BMH to the M3M.
+
+As such, even after deprovisioning we will keep the
+`infrastructure.cluster.x-k8s.io/node-reuse` on a BMH and when M3M will be
+requesting a new BMH, CAPM3 will filter BMH(s) based on their label and will
+pick up those that match the name of KCP/MD.
+
+**Example flow:**
+
+***User updates Metal3MachineTemplate(M3MTemplate):***
+
+1. User has set `NodeReuse` field to True in the new M3MTemplate;
+2. User applies new M3MTemplate (i.e with new image version);
+3. User updates reference in KCP/MD to the newly created M3MTemplate;
+4. Upgrade process starts in KCP/MD;
+5. At this point, KCP/MD triggers a deletion of the old machines which results
+    in the deletion and deprovisioning of machines and BMH(s) respectively by the
+    CAPM3 controller;
+6. While BMH(s) is/are deprovisioning, CAPM3 machine controller will read the
+    `NodeReuse` field value from the new M3MTemplate that KCP is pointing to, and:
+    * If it is set to **True:**
+      * it will set the `infrastructure.cluster.x-k8s.io/node-reuse` label on
+      the BMH, which stores the name of the KCP/MD. BMH becomes Ready again with
+      KCP/MD name set on label and free to be picked up;
+    * If it is set to **False:**
+      * Normal operation continues.
+7. KCP will create new machines, and at the same time new M3M(s) will be created.
+    At this point, CAPM3 Machine controller will look in the newly created M3M(s)
+    for annotation `clonedFrom` which will be always pointing to the M3MTemplate.
+    In the following step, it will find the `NodeReuse` field in the M3MTemplate
+    and:
+    * If it is set to **True:**
+      * CAPM3 Machine controller will filter out ready BMH(s) labelled with
+      `infrastructure.cluster.x-k8s.io/node-reuse` label set to KCP/MD name:
+        * If BMH is found in Ready state:
+          * Pick that BMH and attach it to newly created M3M;
+          * Clean up the `infrastructure.cluster.x-k8s.io/node-reuse` label from
```

Comment:
Why is it required the cleanup? Do you mean that the entire label will be removed or just its value?

Reply:
In case when the user changes from using `nodeReuse` to **not** using `nodeReuse` feature, we were thinking of removing the whole label so that node should be available for other pools. Would that make sense @andfasano ?

Reply:
I'm personally not a big fan of label removals (usually applied in one-shot events), as they could add a time dependency on the flow, but I think it could work fine in this situation